cdap/resource_streaming_program.go

```diff
@@ -0,0 +1,248 @@
+	for {
+		p, err := getProgramStatus(config, statusAddr)
+		if err != nil {
+			return err
+		}
+
+		switch s := p.Status; s {
+		case "PROVISIONING":
+			log.Println("program still in PROVISIONING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		case "STARTING":
+			log.Println("program still in STARTING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		case "STOPPED":
+			log.Println("program still in STOPPED state, waiting 10 seconds. This may occur when redeploying a previously deployed pipeline.")
+			time.Sleep(10 * time.Second)
+		case "RUNNING":
+			log.Println("program successfully reached RUNNING state.")
+			d.SetId(d.Get("app").(string))
+			return nil
+		case "FAILED":
+			return fmt.Errorf("failed to start program in app: %s in state: %s.", d.Get("app"), p.Status)
+		default:
+			log.Println("program still in STARTING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		}
+	}
+
+	return nil
+}
+
+func resourceStreamingProgramRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func getProgramStatus(config *Config, statusAddr string) (p ProgramStatus, err error) {
+	req, err := http.NewRequest(http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	if err := json.Unmarshal(b, &p); err != nil {
+		return p, err
+	}
+
+	return p, nil
+}
+
+func stopProgram(config *Config, stopAddr string) error {
+	req, err := http.NewRequest(http.MethodPost, stopAddr, nil)
+	if err != nil {
+		return err
+	}
+	_, err = httpCall(config.httpClient, req)
+	return err
+}
+
+func resourceStreamingProgramDelete(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+	name := d.Get("name").(string)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		name)
+
+	// Check status to handle scenarios like a program that is in STOPPING state.
+	statusAddr := urlJoin(
+		addr, "/status")
+	stopAddr := urlJoin(
+		addr, "/stop")
+
+	// Poll until actually reaches STOPPED state.
+	for {
+		p, err := getProgramStatus(config, statusAddr)
+		if err != nil {
+			return err
+		}
+
+		switch s := p.Status; s {
+		case "STOPPED":
+			return nil
+		case "RUNNING":
+			err = stopProgram(config, stopAddr)
+			if err != nil {
+				return err
+			}
+		case "STOPPING":
+			log.Println("program still in STOPPING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		default:
+			return fmt.Errorf("cannot stop program in app %s in state: %s", d.Get("app"), p.Status)
+		}
+	}
+}
+
+func resourceStreamingProgramExists(d *schema.ResourceData, m interface{}) (bool, error) {
+	config := m.(*Config)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		d.Get("name").(string), "/status")
```

Comment:
Instead of status this should probably get the list of programs and check for the name